tip-distributor/src/claim_mev_workflow.rs

```diff
@@ -32,121 +39,343 @@ pub enum ClaimMevError {
+        tree_nodes
+            .iter()
+            .map(|tree_node| tree_node.claim_status_pubkey)
+            .collect_vec(),
+    )
+    .await
+    .unwrap();
+    let account_fetch_elapsed = account_fetch_start.elapsed();
+
+    for tree in merkle_trees.generated_merkle_trees {
+        let fetched_tip_distribution_account = match tdas.get(&tree.tip_distribution_account) {
+            Some(account) => account,
+            None => panic!(
+                "TDA not found in cache for account: {:?}",
+                tree.tip_distribution_account
+            ),
+        };
+        // only claim for ones that have merkle root on-chain
+        if fetched_tip_distribution_account.merkle_root.is_none() {
+            info!(
+                "Merkle root has not uploaded yet. Skipped {} claimants for TDA: {:?}",
+                tree.tree_nodes.len(),
+                tree.tip_distribution_account
+            );
+            skipped_merkle_root_count = skipped_merkle_root_count.checked_add(1).unwrap();
+            continue;
         }
-        let stake_acct_min_rent = rpc_client.get_minimum_balance_for_rent_exemption(StakeStateV2::size_of()).await.expect("Failed to calculate min rent");
-        let mut below_min_rent_count: usize = 0;
-        let mut zero_lamports_count: usize = 0;
-        for tree in merkle_trees.generated_merkle_trees {
-            // only claim for ones that have merkle root on-chain
-            let account = rpc_client.get_account(&tree.tip_distribution_account).await.expect("expected to fetch tip distribution account");
-            let fetched_tip_distribution_account = TipDistributionAccount::try_deserialize(&mut account.data.as_slice()).expect("failed to deserialize tip_distribution_account state");
-            if fetched_tip_distribution_account.merkle_root.is_none() {
-                info!(
-                        "not claiming because merkle root isn't uploaded yet. skipped {} claimants for tda: {:?}",
-                        tree.tree_nodes.len(),
-                        tree.tip_distribution_account
-                    );
+        for node in tree.tree_nodes {
+            if node.amount == 0 {
+                zero_lamports_count = zero_lamports_count.checked_add(1).unwrap();
                 continue;
             }
-            for node in tree.tree_nodes {
-                if node.amount == 0 {
-                    zero_lamports_count = zero_lamports_count.checked_add(1).unwrap();
+
+            // make sure not previously claimed
+            match claim_statuses.get(&node.claim_status_pubkey) {
+                Some(Some(_account)) => {
+                    debug!(
+                        "Claim status account already exists (already paid out). Skipping pubkey: {:?}.", node.claim_status_pubkey,
+                    );
+                    already_claimed_count = already_claimed_count.checked_add(1).unwrap();
                     continue;
                 }
+                None => panic!(
```

Comment:
can you document why this would happen? preferably would like to avoid panicking

Reply:
seems like would happen if rpc error, in that case we should fail before we get to this step

Reply:
yes, this unrap() should prevent this case: https://github.com/jito-foundation/jito-solana/pull/392/files#diff-f2fd8703d48635e7d3a0ec5182e43a1dde5ae307fa6f637d2025162d692c24b3R162